src/what-unsafe-does.md

```diff
@@ -16,16 +16,19 @@ to your program. You definitely *should not* invoke Undefined Behavior.
 Unlike C, Undefined Behavior is pretty limited in scope in Rust. All the core
 language cares about is preventing the following things:
 
-* Dereferencing null, dangling, or unaligned pointers
+* Dereferencing null, dangling, or unaligned references or raw pointers
+* Performing out-of-bounds arithmetic for the computation of a struct/tuple
+  field address
```

Comment:
I don't know what you are asking. How is that UB, or how does one do the arithmetic?

Reply:
Yeah I can't see how to do this in a way that *wouldn't* also apply to safe references but isn't just a call to an unsafe library function (which all define their own UB contracts and aren't in scope for this page).

Reply:
That's a good point if I understand your meaning: "references" doesn't belong on line 19 because they already can't be null, dangling, or unaligned.

Reply:
Right yeah, I think this was written before we came to the conclusion that the *existence* of bad references was immediate UB (the golden days of access-based validity....)

Reply:
oh no wait fuck ralf added the references part brains pls 😖

Reply:
> (the golden days of access-based validity....)

You mean when what we documented as UB and what we compiled for LLVM were not matching? ;)

Indeed for references I *think* this is somewhat overlapping with the rule that you cannot even create dangling references, but for raw pointers it is not. I am also not sure about references to unsized types.

Reply:
So do you want me to change something here now?

Reply:
I would drop mention of references here (reverting to the checked-in state); if you're not allowed to create bad ones, using bad ones is irrelevant.

And I would completely drop any mention of offsets unless you can provide a case where this happens with raw pointers and no other rule covers and it's *not* about using offset/add (they document it, and wrapping_offset specifically allows it, so it is in fact not intrinsic to the language).

Are you picturing something like creating a type whose `size_of` is `> usize::MAX` (does the compiler even permit this??), and relying on a field access to wrap around the address space and alias some earlier part of the type???

Reply:
I think you could do something like this:
```rust
fn main() {
    let local = 5u8;
    let ptr = (&local as *const u8).wrapping_sub(1) as *const (u8, u8);
    let _ref = unsafe { &(*ptr).1 };
}
```
This should be UB but I don't think it violates any of the other clauses?

Also with https://github.com/rust-lang/rfcs/pull/2582 it is easy to construct an example:
```rust
fn main() {
    let ptr: *const (u8, u8) = 16usize as *const _;
    let ptr2 = &raw const (*ptr).1;
}
```

Reply:
But I guess this relates to the other discussion -- they might both be UB because a dangling raw ptr gets "dereferenced" even though no access happens. I thought it was established that we wanted to require as little as possible for raw pointers.

Funny how I am arguing for an access-based model now and you are arguing against it... ;)

Reply:
Doesn't the fact that it does wrapping_sub to go out of bounds of the allocated object make it UB to use that pointer?

I mean I know that wrapping_sub allows the math to go out of bounds without being UB, but the resulting pointer is invalid if out of bounds, right?

Reply:
Using `wrapping_add` it is certainly possible to make this a valid pointer again.

Reply:
Uh, right. "use" is too colloquial. "access through" is the term I suppose.

Reply:
At least in my dictionary, the two examples above however don't "access" through the raw ptr. They do a ptr-to-place conversion (that's the `*` operator), then restrict the place to a field, and then a place-to-ptr coercion (the `&` operator). No memory access.

Reply:
hmmm, this is a tricky one. I'll try and sit down at my desktop and do the heavy thinking thing later today.